### DISCUSSION TITLE
Eventually, the Node ecosystem has to move to ESM

### DISCUSSION THREAD

**Discussion by thetutlage** [adonisjs/core#3689] [Polls]
Node.js has supported ES modules as a first-class citizen for a while now. As a result, you can use them without any transpiler or bundler.

However, moving from CommonJS `(require('some-package'))` to ESM `import 'some-package'` is a breaking change at many levels. I will not go in-depth about those breaking changes right now but leave some pointers that you help you make your decision for the vote.

#### All relative imports needs a `.js` extension. 

```js
// ❌ Does not work
import User from './user'

// ✅ Works
import User from './user.js'
```

#### Directory imports with `index.js` no longer work

```ts
// ❌ Does not work
import * as components from './components'

// ✅ Works
import * as components from './components/index.js'
```

#### Globals like `__filename`, `__dirname`, and `require` are not available
As the title says, you cannot use `__filename`, `__dirname` globals inside your code, and neither use methods like `resolve` from the `require` object.

#### JSON imports needs assertion types
You cannot import a JSON file similar to a JavaScript file (earlier, it was possible). Instead, you need to use the following syntax.

```json
import pkg from './package.json' assert { type: 'json' };
```

## Can we live without migrating to ESM?
Well, not in the longer run. A lot of packages are already moving to ESM only. For example, a prolific Node.js developer [Sindresorhus has decided to move all his packages to ESM](https://github.com/sindresorhus/meta/discussions/15), and therefore, the core of AdonisJS + your application will not be able to use his packages (and any other ESM only package).

Therefore, the change is inevitable, and sooner or later, we will have to migrate.

## Can migration be automated?
To a certain level, yes. I am playing with [codemods](https://github.com/facebookarchive/codemod), which can rewrite some of the imports automatically for you. However, not everything can be automated and will require manual effort (at least for verification)

**Comment by ndianabasi** [comment]
Everything is in a state of flux. Only those who are adaptable can survive. We love AdonisJS. 💖💖

**Comment by brunoalfred** [comment]
I think it's a good idea to move with the trend instead of being left behind. AdonisJS ❤️

**Comment by zakariamehbi** [comment]
Why not, the sooner the better.

**Comment by jacobq** [comment]
This is an easy 👍 vote for me because I am not doing the work 🤣. I believe you are quite right in that this is the direction active node packages should move as it is newer, broader (works in browsers!), and actually part of the ECMAScript standard. That said, there is a trade-off: do it earlier and enjoy the benefits vs. wait longer and enjoy a smoother transition thanks to the work of others who have gone before (and maybe written a few codemods, blog posts, etc.).

FWIW, you can mimic the functionality of `__dirname` and `__filename` [like this](https://leveluptutorials.com/posts/how-to-use-__dirname-with-esm):
```js
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
```

You can even get `require` back if you need it via [`createRequire`](https://nodejs.org/dist/latest-v18.x/docs/api/module.html#modulecreaterequirefilename).

**Comment by muco-rolle** [comment]
My vote would have been **No** but It seems like it's inevitable.

**Comment by Twitch0125** [comment]
Would it be possible to make Adonis compatible with Deno at the same time? I've seen other node projects (like Nuxt 3) become "runtime agnostic" to where they can run in Deno, Node, or the Browser and I think that's another part of the Future Proofing question

**Comment by redeemefy** [comment]
What is the NodeJS recommendation about `__dirname` and `__filename`? How NodeJS itself is using/solving for globals?

If Adonis move forward with the language, that in my opinion it should, it should explain/recommend the path forward on how to implement those breaking changes.

```typescript
// before
path.join(__dirname, ...)

// after
...
```

**Reply by thetutlage** [reply]
Yeah, there will be documentation around migration. This post was to understand if most of the community is in favor of moving to ESM, knowing that fact there will be breaking changes

**Comment by dipanjanpanja6** [comment]
if adonis 5 had a sable version then there is no problem.

**Reply by thetutlage** [reply]
AdonisJS is stable for over a year now :)

**Comment by SH8GH** [comment]
if i can suggest, i think the `@ioc` concept new syntax.

example:

```ts
import View from '@ioc:Adonis/Core/View' => const View = import.meta.container('@ioc:Adonis/Core/View')
```
